prover/crates/bin/prover_autoscaler/README.md

```diff
@@ -0,0 +1,231 @@
+export metrics (path is `/metrics`), and `http_port` with 3 paths: `/healthz`, `/cluster` to get the cluster status and
+`/scale` to scale Deployments up or down.
+
+### Scaler
+
+Scaler collects cluster statuses from Agents, job queues from prover-job-monitor, calculates needed number of replicas
+and sends scale requests to Agents.
+
+Requests flow diagram:
+
+```mermaid
+sequenceDiagram
+    participant prover-job-monitor
+    participant Scaler
+    box cluster1
+        participant Agent1
+        participant K8s API1
+    end
+    box cluster2
+        participant Agent2
+        participant K8s API2
+    end
+    loop Watch
+        Agent1->>K8s API1: Watch namespaces
+    end
+    loop Watch
+        Agent2->>K8s API2: Watch namespaces
+    end
+    loop Recalculate
+        Scaler->>prover-job-monitor: /report
+        Scaler->>Agent1: /cluster
+        Scaler->>Agent2: /cluster
+        Scaler->>Agent1: /scale
+    end
+```
+
+Scaler supports 2 types of scaling algorithms: GPU and Simple. GPU usually is prover itself and all other Deployments
+are using Simple algorithm.
+
+Simple algorithm tries to scale the Deployment up too `queue / speed` replicas (rounded up) in the best cluster. If
+there is not enough capacity it continue in the next best and so on. On each run it selects "best cluster" using
+priority, number of capacity issues and cluster size.
+
+GPU algorithm works similar to Simple one, but it also recognise different GPU types and distribute load across L4 GPUs
+first, than T4, V100, P100 and A100, if available.
+
+Different namespaces are running different protocol versions and completely independent. Normally only one namespace is
+active, and only during protocol upgrade both are active. Each namespace has to have correct version of binaries
+installed, see `protocol_versions` config option.
+
+## Dependencies
+
+- [prover-job-monitor](.../prover_job_monitor/)
+- Kubernetes API
+- GCP API (optional)
+
+## Permissions
+
+Agents need the following Kubernetes permissions:
+
+```yaml
+- apiGroups:
+    - ''
+  resources:
+    - pods
+    - events
+    - namespaces
+    - nodes
+  verbs:
+    - get
+    - watch
+    - list
+- apiGroups:
+    - apps
+  resources:
+    - deployments
+    - replicasets
+  verbs:
+    - get
+    - list
+    - watch
+    - patch
+    - update
+```
+
+## Configuration
+
+Prover Autoscaler requires a config file provided via `--config-path` flag, supported format: YAML. Also you need to
+specify which job to run Scaler or Agent using `--job=scaler` or `--job=agent` flag correspondingly.
+
+### Common configuration
+
+- `graceful_shutdown_timeout` is time to wait for all the task to finish before force shutdown. Default: 5s.
+- `observability` section configures type of `log_format` (`plain` or `json`) and log levels per module with
+  `log_directives`.
+
+Example:
+
+```yaml
+graceful_shutdown_timeout: 5s
```

Comment:
Using `..._secs` makes the config value less readable for humans (you have to read the name, split it into actual one and type, then read value, recall type and convert for example 600 to 10min) and for the code: you have to somewhere convert from integer into Duration.
Also fixing type as seconds prevents people from specifying 0.5s for example.

Using here Duration eliminates need for custom conversion, easier to read and more flexible. `ms` or `ns` will be quite visible during review, but if it will be an issue  we can start validate configuration tightly, like specifying min and max values.